### ISSUE TITLE
Ubuntu 19.10/20.04 build - disable LTO required to avoid internal compiler error

### ISSUE THREAD

**Issue by hmmbug** [christofmuc/KnobKraft-orm#6]
Hi,

Thanks for all your work on this. From the short time I've had using it, I think it's going to be a really useful tool for my Rev2. I encountered a small problem building on Linux though.

On Ubuntu 19.10 Eoan, when building from commit f83126864441ec116aefcd42a6f1aaa81dfb0411 I get the following error:

```
...snip...
[ 78%] Linking CXX shared module pytschirp.cpython-37m-x86_64-linux-gnu.so
lto1: internal compiler error: in add_symbol_to_partition_1, at lto/lto-partition.c:153
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-9/README.Bugs> for instructions.
lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[2]: *** [pytschirp/CMakeFiles/pytschirp.dir/build.make:232: pytschirp/pytschirp.cpython-37m-x86_64-linux-gnu.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:1021: pytschirp/CMakeFiles/pytschirp.dir/all] Error 2
make: *** [Makefile:133: all] Error 2
```
I can work around it prefixing `LDFLAGS="--disable-lto"` to cmake:

```
LDFLAGS="--disable-lto" cmake -S . -B builds && cmake --build builds
```
Unfortunately I don't have enough expertise with these build tools to figure out a permanent fix. I'm happy to do some testing for you if that helps. I'm using cmake 3.16.6 (self compiled as the distro version is too old at 3.13.4) and c++ 9.2.1 (distro package). 

Thanks!

**Comment by christofmuc** [comment]
Thank you very much! I had run into that exact problem on my Debian 9 with gcc 8.3.0, and was not able to solve it!

It seems to be a regression somewhere in gcc, as it works on an older gcc 7.5.0. The culprit is pybind11 used for the python integration, and it forces LTO on Linux unless you globally disable it. 

I found a way to make pybind11 behave, even if that requires insight into the CMake file of pybind11, but for me it works. If you like, you can try out commit #180e6cd8e26109f2cd24516afbf17ad26811a5fb which contains my version of the fix.

Again, many thanks for contributing and I am glad you like the tool . If you have any ideas or features wishes, let me know!

**Comment by hmmbug** [comment]
Many thanks - commit 180e6cd8e26109f2cd24516afbf17ad26811a5fb builds without needing the additional LDFLAGS shell variable.

**Comment by christofmuc** [comment]
Awesome, thanks for the confirmation!

**Comment by christofmuc** [comment]
Somehow this problem is back on my freshly installed Ubuntu 20.04 LTS with gcc 9.3.0, only that the fix from last year no longer works. This is strange, I cannot recall that I updated the pybind11 library but somehow that seems the most likely explanation. I will need to dig deeper here.

**Comment by hmmbug** [comment]
FYI, I tested these build ok out-of-the-box:

- Ubuntu 20.10 (gcc/c++ 10.3.0-1ubuntu1~20.10)
- Ubuntu 21.10 (gcc/c++ 11.2.0-7ubuntu2)

Upgrading to g++ 10 on 20.04 (```sudo apt install gcc-10 g++-10```) still has the LTO errors at linker stage, perhaps it's related to the linker version in the binutils 2.34 package. Ubuntu 20.10 and 21.10 have newer versions (2.35 and 2.37 respectively) but there doesn't seem to be a simple (eg. apt install) method to upgrade it so I didn't try further.

Back to 20.04 with gcc/c++ 9.3.0... I tried a few different LTO-related LDFLAGS (namely ```--disable-lto```, ```-fno-lto```, and ```-flto=thin```) with no success. The ```CMAKE_INTERPROCEDURAL_OPTIMIZATION``` switch (mentioned in the pybind11 docs) seems to work though - it compiles ok and runs... 
```
cmake -D CMAKE_INTERPROCEDURAL_OPTIMIZATION=off -S . -B builds
cmake --build builds
```
Note though, I did this test quickly in a temporary VM - my workstation is now running 21.10 so I can't give it a proper test. I hope this helps a bit though.
